File: 第二章_排序/排序(3)/重复值判断练习题/Checker.cpp
class Checker {
public:
    bool checkDuplicate(vector<int> a, int n) {
       a=headSort(a,n);
	   int i;
	   for(i=0;i<a.size()-1;i++){//排好序的数组若相邻的数相等则代表有重复的值出现
		   if(a[i]==a[i+1]){
			  return true;
		   }
	   }

	   return false;
    }
	
	//堆排序
	vector<int> headSort(vector<int> a,int n){
		int i;

		for(i=n/2;i>=0;i--){
			head_and_just(a,i,n-1);
		}

		for(i=n-1;i>=0;i--){
			swap(a,0,i);
			head_and_just(a,0,i);
		}

		return a;
	}


	void head_and_just(vector<int> &A,int parent,int length){
		int temp=A[parent];
		int child=parent*2+1;

		while(child<length){
			if(child+1<length&&A[child+1]>A[child]){
				child++;
			}

			if(temp>A[child]){
				break;
			}

			A[parent]=A[child];
			parent=child;
			child=child*2+1;
		}

		A[parent]=temp;
	}

	void swap(vector<int> &A,int a,int b){
		int t=A[a];
		A[a]=A[b];
		A[b]=t;
	}
};

File: 第二章_排序/排序(3)/重复值判断练习题/Checker.java
import java.util.*;
 
public class Checker {
    public boolean checkDuplicate(int[] a, int n) {
        int[] ints = HeadSort(a, n);
        for (int i=1;i<n;i++){
            if (ints[i]==ints[i-1]){
                return true;
            }
        }
        return false;
    }
    public int [] HeadSort(int [] A,int n){
 
        for (int i=n/2;i>=0;i--){
            HeadAdjust(A,i,n);
        }
        for (int i=n-1;i>=0;i--){
            int temp=A[0];
            A[0]=A[i];
            A[i]=temp;
            HeadAdjust(A,0,i);
        }
        return A;
    }
 
    /**
     * ½¨Á¢´ó¶¥¶Ñ
     * @param A
     * @param parent
     * @param length
     */
    public void HeadAdjust(int A[],int parent,int length){
        int temp=A[parent];
        int child=parent*2+1;
        while (child<length){
            if (child+1<length&&A[child]<A[child+1]){
                child++;
            }
            if (temp>=A[child]){
                break;
            }
            A[parent]=A[child];
            parent=child;
            child=child*2+1;
        }
        A[parent]=temp;
    }
}